File: packages/paso-py/usepaso/utils/redact.py
"""Redact sensitive query parameter values from URLs for safe logging."""

from urllib.parse import urlparse, parse_qsl, urlunparse

SENSITIVE_PARAMS = frozenset({
    'api_key', 'apikey', 'api-key',
    'token', 'access_token', 'auth_token',
    'secret', 'client_secret',
    'password', 'passwd',
    'key',
})


def redact_url(url: str) -> str:
    """Replace values of sensitive query params with '***'."""
    try:
        parsed = urlparse(url)
        if not parsed.query:
            return url
        pairs = parse_qsl(parsed.query, keep_blank_values=True)
        redacted = False
        new_pairs = []
        for key, value in pairs:
            if key.lower() in SENSITIVE_PARAMS:
                new_pairs.append(f'{key}=***')
                redacted = True
            else:
                new_pairs.append(f'{key}={value}')
        if not redacted:
            return url
        new_query = '&'.join(new_pairs)
        return urlunparse(parsed._replace(query=new_query))
    except Exception:
        return url

File: packages/paso-js/src/utils/redact.ts
/**
 * Redact sensitive query parameter values from URLs for safe logging.
 * Matches common parameter names used for authentication.
 */

const SENSITIVE_PARAMS = new Set([
  'api_key',
  'apikey',
  'api-key',
  'token',
  'access_token',
  'auth_token',
  'secret',
  'client_secret',
  'password',
  'passwd',
  'key',
]);

export function redactUrl(url: string): string {
  try {
    const parsed = new URL(url);
    let redacted = false;
    for (const [key] of parsed.searchParams) {
      if (SENSITIVE_PARAMS.has(key.toLowerCase())) {
        parsed.searchParams.set(key, '***');
        redacted = true;
      }
    }
    return redacted ? parsed.toString() : url;
  } catch {
    return url;
  }
}
